The Block Clock

johnny9 opened the pull request:
This is a continuation of #148. The goal of this PR is to get the working features from #148 merged in using a QQuickPainterItem implementation of the dial instead of the javascript canvas.  Connection state as well as the animating icon will be implemented in a later PR. Refactoring and clean up from the removal of the Canvas code still remains.

[![Windows](https://img.shields.io/badge/OS-Windows-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/win64/insecure_win_gui.zip?branch=pull/220)
[![Intel macOS](https://img.shields.io/badge/OS-Intel%20macOS-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos/insecure_mac_gui.zip?branch=pull/220)
[![Apple Silicon macOS](https://img.shields.io/badge/OS-Apple%20Silicon%20macOS-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos_arm64/insecure_mac_arm64_gui.zip?branch=pull/220)
[![ARM64 Android](https://img.shields.io/badge/OS-Android-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/android/insecure_android_apk.zip?branch=pull/220)

johnny9 commented:
Update from[ 90469cb ](https://github.com/bitcoin-core/gui-qml/commit/90469cb85211ecf3933602d33712eb95c89f2838)to[ e9b5dd](https://github.com/bitcoin-core/gui-qml/commit/e9b5ddd27596b647f7f56ef72302bcba3771a9e8):
 - Replace tabs with spaces

johnny9 commented:
Update branch from[ e9b5ddd ](https://github.com/bitcoin-core/gui-qml/commit/e9b5ddd27596b647f7f56ef72302bcba3771a9e8)to 96e47f8:
 - Fix painting of last block. index 0 of timeRatioList is actually the current time.

johnny9 commented:
Update from[ 96e47f8 ](https://github.com/bitcoin-core/gui-qml/commit/96e47f8242c6be0dce8e4235bfdff483d4fdf586)to[ 63792bf](https://github.com/bitcoin-core/gui-qml/commit/63792bfca32b3327aecd0b9577c24fd0f175ccef):
  - Fix Makefile includes

johnny9 commented:
Update branch from[ 37b6257 ](https://github.com/bitcoin-core/gui-qml/commit/37b6257bd3f4951d18cb323a8fc06b8423a0a211)to 8515fdf:
  - Added comment for calculating the "gap" between painted blocks.

johnny9 commented:
I noticed that these changes require the QtGraphicalEffects module (qml-module-graphicaleffects ubuntu package). I'll likely have to update the depends to fix static builds or change how the icon is forced into a specific color.

johnny9 commented on src/qml/chainmodel.cpp:
```diff
@@ -0,0 +1,80 @@
+    double current_time_ratio = double(secs_since_meridian) / SECS_IN_12_HOURS;
+
+    if (current_time_ratio < m_time_ratio_list[0].toDouble()) { // That means time has crossed a meridian
+        m_time_ratio_list.erase(m_time_ratio_list.begin() + 2, m_time_ratio_list.end());
```
This line causes an abort during sync.  Need to root cause and fix before merge.

johnny9 replied:
Ok, I think i see it. During sync there is likely no blocks in the list so erase call in that range is invalid.

jarolrod reviewed (commented):
concept ack

jarolrod reviewed (commented):
Let's squash some of the history to have a clean commit record, but also I'm seeing some slight misalignment with the design file. 

Below, the more visible layer is the design file, and below that is this PR on signet (hence diff block heights). Looks like the block height, the words "Blocktime" and peers indicator should be a bit higher up. Also fine with leaving this fixup for a follow-up!

<img width="786" alt="Screen Shot 2023-01-25 at 12 05 42 AM" src="https://user-images.githubusercontent.com/23396902/214484384-fa12d111-5441-490a-ba45-08aaa7380238.png">

johnny9 commented:
Update branch from[ 8ecfae6 ](https://github.com/bitcoin-core/gui-qml/commit/8ecfae6227047c66242881c4b50d5527bbf5fc3f)to 76a504e:
 - Squashed all BlockClock control commits into one

johnny9 commented:
Update branch from[ 76a504e ](https://github.com/bitcoin-core/gui-qml/commit/76a504e826482a5ff62cffeea4ec9863d2a746d9)to[ 2e5a0d8](https://github.com/bitcoin-core/gui-qml/commit/2e5a0d83c2bcfdfc878a46b2ce1d23d1ceb74b8d):
  - Adjusted alignment of labels to better match the figma spec
  
  
![Screenshot from 2023-01-28 10-37-08](https://user-images.githubusercontent.com/985648/215276307-a1675e9e-de17-4aec-88fd-f909fa6eddff.png)

johnny9 commented:
Update. Remove a missed trailing whitespace

johnny9 commented:
Update from[ 3fd696d ](https://github.com/bitcoin-core/gui-qml/commit/3fd696d6036fe737bf1c9dbb8427cc838d4dda7a)to 33eb4af: 
 - Fixed potential initialization error in ChainModel, possible root cause of an issue Jarol discovered with the dial not lining up with 12 o'clock.

jarolrod commented on src/qml/components/blockclockdial.h:
```diff
@@ -0,0 +1,56 @@
+    void paintBackground(QPainter * painter);
+    void paintTimeTicks(QPainter * painter);
+    QRectF getBoundsForPen(const QPen & pen);
+    double degrees_per_pixel();
```
Per [developer notes](https://github.com/bitcoin-core/gui-qml/blob/main/doc/developer-notes.md#coding-style-c)


> Class names, function names, and method names are UpperCamelCase (PascalCase). Do not prefix class names with C. See Internal interface naming style for an exception to this convention.

johnny9 replied:
Thanks, will fix this.

jarolrod commented on src/qml/controls/BlockClock.qml:
```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2023 The Bitcoin Core developers
```
Why should this be a control, and the block clock dial a component?

johnny9 replied:
This item can take focus and accept input.  The dial just renders the blocks.

johnny9 replied:
Actually not really convinced of this. Can see   it living in components and making sense too

johnny9 replied:
Really liked how you structured the commits in that branch. Going to use that.

jarolrod reviewed (commented):
Almost there, just one change needed. 

Additionally, i'd structure the commits something similar to this: https://github.com/jarolrod/gui-qml/tree/block-clock-review

But no necessary.

johnny9 commented:
Update branch to 6701ad2:
 - Fixed code style issue in blockclock
 - Moved BlockClock.qml to components
 - Split BlockClock commit up to clarify history

jarolrod commented on src/qml/components/BlockClock.qml:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    MouseArea {
+        anchors.fill: dial
```
could add, since the paused state is implemented
```suggestion
        anchors.fill: dial
        cursorShape: Qt.PointingHandCursor
```

jarolrod commented on src/qml/nodemodel.cpp:
```diff
@@ -27,14 +28,60 @@ void NodeModel::setBlockTipHeight(int new_height)
     }
 }
 
+void NodeModel::setRemainingSyncTime(double new_progress)
```
This function is pretty confusing in general, we'll heavily change this in follow-ups

jarolrod replied:
should probably do an early test on new_progress to make sure it's a valid value

jarolrod commented on src/qml/nodemodel.cpp:
```diff
@@ -27,14 +28,60 @@ void NodeModel::setBlockTipHeight(int new_height)
+        int timeDelta = 0;
+        int remainingMSecs = 0;
+        double remainingProgress = 1.0 - new_progress;
+        for (int i = 1; i < m_block_process_time.size(); i++) {
```
starting from 1, isn't it skipping the first value in the list?

jarolrod commented on src/qml/nodemodel.cpp:
```diff
@@ -27,14 +28,60 @@ void NodeModel::setBlockTipHeight(int new_height)
+        for (int i = 1; i < m_block_process_time.size(); i++) {
+            QPair<int, double> sample = m_block_process_time[i];
+
+            // take first sample after 500 seconds or last available one
```
but `currentTime` was set in milliseconds in line 33, now we're doing math with seconds values?

jarolrod commented on src/qml/nodemodel.cpp:
```diff
@@ -27,14 +28,60 @@ void NodeModel::setBlockTipHeight(int new_height)
+        }
+        static const int MAX_SAMPLES = 5000;
+        if (m_block_process_time.count() > MAX_SAMPLES) {
+            m_block_process_time.remove(1, m_block_process_time.count() - 1);
```
not sure why we're removing the first element here; but in follow-ups its better if this is a queue, instead of a vector that we manually remove elements from

jarolrod reviewed (approved):
ACK 6701ad24382ae4a23d3e55c8d952df2ca061e737

This works, with improvements to be made in follow-ups
